system/security/LdapSecurity/ldapconnection.cpp

```diff
@@ -6163,7 +6163,63 @@ class CLdapClient : implements ILdapClient, public CInterface
 
     virtual aindex_t getManagedFileScopes(IArrayOf<ISecResource>& scopes)
     {
-        getResourcesEx(RT_FILE_SCOPE, m_ldapconfig->getResourceBasedn(RT_FILE_SCOPE), NULL, NULL, scopes);
+        //Get array of all file scopes listed in files baseDN
+        StringBuffer basednbuf;
+        LdapUtils::normalizeDn(m_ldapconfig->getResourceBasedn(RT_FILE_SCOPE), m_ldapconfig->getBasedn(), basednbuf);
+        basednbuf.toLowerCase();//will look something like "ou=files,ou=dataland_ecl,dc=internal,dc=sds"
+
+        TIMEVAL timeOut = {m_ldapconfig->getLdapTimeout(),0};
+        Owned<ILdapConnection> lconn = m_connections->getConnection();
+        LDAP* ld = ((CLdapConnection*)lconn.get())->getLd();
+        char *attrs[] = {"canonicalName", NULL};
+
+        //Call LDAP to get the complete OU tree underneath basdnbuf
+        CPagedLDAPSearch pagedSrch(ld, m_ldapconfig->getLdapTimeout(), (char*)basednbuf.str(), LDAP_SCOPE_SUBTREE, "objectClass=*", attrs);
+        for (LDAPMessage *message = pagedSrch.getFirstEntry(); message; message = pagedSrch.getNextEntry())
+        {
+            CLDAPGetAttributesWrapper   atts(ld, message);
+            for ( char *attribute = atts.getFirst();
+                  attribute != NULL;
+                  attribute = atts.getNext())
+            {
+                CLDAPGetValuesLenWrapper vals(ld, message, attribute);
+                if (vals.hasValues())
+                {
+                    const char* val = vals.queryCharValue(0);
+                    if(val)//this check probably isn't necessary since hasValues(), but could prevent a core
+                    {
+                        //Build filescope from everything below the file scope basedn
+                        StringArray OUarray;
+                        OUarray.appendList(val,"/",true);//create StringArray of OU elements. LDAP returns them in the form "internal.sds/files/dataland_ecl/hpccinternal/roxieuser
+
+                        //Find index of first StringArray item NOT in base search string (basednbuf)
+                        int finger = 1;//skip the first element, which are the combined domains. So "dc=internal,dc=sds" will appear here as "internal.sds"
+                        int end = OUarray.ordinality();
+                        for (; finger < end; finger++)
+                        {
+                            VStringBuffer theOU("ou=%s,",OUarray.item(finger));
+                            theOU.toLowerCase();
+                            if (nullptr == strstr(basednbuf.str(), theOU.str()))//search baseDN for OU substring
```

Comment:
Because I add "ou=" to the beginning of each token, and a trailing comma, this will not be a false positive
ou=file,
ou=fi, 
ou=les,

Reply:
yes, ok.

Reply:
can you guarantee that the last ou in basednbuff has a trailing comma?  Could the dc= come first?  If not this might fail to match.

Reply:
Yes, I can guarantee it.  The basednbuff is combined from the DC, which is queried from LDAP (dc=internal,dc=sds), and the filesBaseDN (ou=files,ou=dataland_ecl) which comes from the environment file . Once scrubbed it will look like ou=files,ou=dataland_ecl,dc=internal,dc=sds.  So all the OUs come first, ensuring a trailing comma. I have tested this trying to break it with no luck so far